tls/s2n_cipher_suites.c

```diff
@@ -1209,13 +1242,33 @@ static int s2n_set_cipher_as_server(struct s2n_connection *conn, uint8_t *wire,
     const struct s2n_security_policy *security_policy;
     POSIX_GUARD(s2n_connection_get_security_policy(conn, &security_policy));
 
-    /* s2n supports only server order */
-    for (int i = 0; i < security_policy->cipher_preferences->count; i++) {
-        const uint8_t *ours = security_policy->cipher_preferences->suites[i]->iana_value;
+    const struct s2n_cipher_preferences *cipher_preferences = security_policy->cipher_preferences;
+    POSIX_ENSURE_REF(cipher_preferences);
 
-        if (s2n_wire_ciphers_contain(ours, wire, count, cipher_suite_len)) {
+    bool client_signalled_chacha20_boosting = false;
+    if (cipher_preferences->allow_chacha20_boosting) {
+        POSIX_GUARD_RESULT(s2n_client_signalled_chacha20_boosting(wire_stuffer, cipher_suite_len, &client_signalled_chacha20_boosting));
+    }
+
+    /* 
+     * The server will attempt to negotiate a chacha20 cipher suite if both the allow_chacha20_boosting flag is enabled  
+     * AND the client has a chacha20 cipher suite first in its cipher preference list
+     */
+    bool mutual_chacha20_boosting_support = cipher_preferences->allow_chacha20_boosting && client_signalled_chacha20_boosting;
```

Comment:
This logic belongs in s2n_client_signalled_chacha20_boosting (which would need to be renamed), not in s2n_set_cipher_as_server. s2n_set_cipher_as_server just needs to know whether or not to do chacha20 boosting. It doesn't care whether it can't because the cipher preferences don't allow it or because the client doesn't signal it might be beneficial.

You should set one bool, something like "try_chacha20_boosting"